File: 백준/35293_멘헤라/Main.java
import java.io.*;
import java.util.*;

/**
 * N이 주어진다, N은 10^16 + 1 이하의 실수이며, 소수점 첫자리 까지 주어진다. 
 * 취하는 정도는 9.0, 7.0 , 4.5, -2.0 이다.
 * N을 달성하기 위한 최소한의 섭취 횟수를 구해야한다. 
 * 부동소수점 계산 오차를 줄이기 위해 정수로 변환하여 계산한다. 모두 10을 곱한다. 
 */
class Solution {

    private final BufferedReader reader;
    private final BufferedWriter writer;

    long[] al = { 90L, 70L, 45L, -20L }; // -> 18,14,9,-4

    private long N;

    private long GCD = 5L;

    private long MOD = 4L;

    public Solution(BufferedReader reader, BufferedWriter writer) throws IOException {
        this.reader = reader;
        this.writer = writer;
        String s = reader.readLine();
        s = s.replace(".", "");
        N = Long.parseLong(s);

        for (int i = 0 ; i < al.length ; ++i ) {
            al[i] /= GCD;
        }
    }

    private void solve() throws IOException {

        // 5로 나누어 떨어지지 않으면 애초에 도달 불가능
        if (N % GCD != 0) {
            writer.write("-1");
            return;
        }

        long target = N / GCD;

        int LIMIT = 500;

        long[] dist = new long[LIMIT + 1];
        Arrays.fill(dist, -1);

        Queue<Integer> q = new ArrayDeque<>();

        q.add(0);
        dist[0] = 0;

        while (!q.isEmpty()) {

            int v = q.poll();

            for (long mv : al) {

                long nv = v + mv;

                if (nv < 0 || nv > LIMIT) continue;
                if (dist[(int) nv] != -1) continue;

                dist[(int) nv] = dist[v] + 1;
                q.add((int) nv);
            }
        }

        long ans = Long.MAX_VALUE;

        for (int v = 0; v <= LIMIT; v++) {

            if (dist[v] == -1) continue;
            if (v > target) continue;

            long diff = target - v;

            if (diff % 18 != 0) continue;

            ans = Math.min(ans, dist[v] + diff / 18);
        }

        if (ans == Long.MAX_VALUE) writer.write("-1");
        else writer.write(String.valueOf(ans));
    }

    public void run() throws IOException {
        solve();
        writer.newLine();
        writer.flush();
    }
}

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(System.out));
        new Solution(reader, writer).run();
        reader.close();
        writer.close();
    }
}

File: 백준/35293_멘헤라/main.kt

import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.util.LinkedList
import java.util.Queue
import kotlin.math.min


fun main(args: Array<String>) {
    val reader = BufferedReader(InputStreamReader(System.`in`))
    val writer = BufferedWriter(OutputStreamWriter(System.`out`))

    class Solution {
        
        private val alchols = listOf(18L, 14L, 9L, -4L)

        private val GCD = 5

        private val LCM = 126

        private var N: Long = 0L

        init {
            val input = reader.readLine().replace(".", "")
            N = input.toLong()
        }

        fun solve(): String {
            var answer = Long.MAX_VALUE
            if ( N % GCD != 0L ) {
                return "-1"
            }
            N /= GCD

            val q: Queue<Long> = LinkedList<Long>()
            val LIMIT = LCM
            // val LIMIT: Int = 500
            val dist: LongArray = LongArray(LIMIT+1) { -1L }
            dist.fill(-1L)
            dist[0] = 0L

            q.offer(0L)


            while ( q.isNotEmpty() ) {
                val current = q.poll()

                for ( i in 0 until 4 ) {
                    val next = current + alchols[i]

                    if ( next < 0L || next > LIMIT ) {
                        continue
                    }

                    if ( dist[next.toInt()] != -1L ) {
                        continue
                    }
                    dist[next.toInt()] = dist[current.toInt()] + 1
                    q.offer(next)
                }
            }

            for ( v in 0 .. LIMIT ) {
                if ( dist[v.toInt()] == -1L ) continue
                if ( v > N ) continue
                val diff = N - v

                if ( diff % 18 != 0L ) continue
                answer = min(answer, dist[v.toInt()] + diff / 18)
            }

            return if ( answer == Long.MAX_VALUE ) {
                "-1"
            } else {
                answer.toString()
            }
        }

        fun run() {
            val answer = solve()
            writer.write("$answer\n")
            writer.flush()
        }
    }

    val solution = Solution()
    solution.run()
}

